File: packages/sdk-node/src/runs.ts
import type { HttpClient } from "./http.js";
import { AgentjailError } from "./http.js";
import type {
  ExecResult,
  ForkRequest,
  ForkResult,
  RunRequest,
  StreamEvent,
} from "./types.js";
import { encodeExecOptions } from "./exec_options.js";

/** One-shot code execution. */
export class Runs {
  constructor(private readonly http: HttpClient) {}

  /** Run code in a fresh jail and return the result. */
  async create(params: RunRequest): Promise<ExecResult> {
    const body: Record<string, unknown> = { code: params.code };
    if (params.language) body.language = params.language;
    if (params.timeoutSecs !== undefined) body.timeout_secs = params.timeoutSecs;
    if (params.memoryMb !== undefined)    body.memory_mb    = params.memoryMb;
    encodeExecOptions(body, params);
    return this.http.request<ExecResult>({
      method: "POST",
      path: "/v1/runs",
      body,
    });
  }

  /**
   * Live-fork: spawn a parent, mid-run COW-clone its output, then spawn
   * one or more children against the forked filesystem state. Pass either
   * `childCode` (single child) or `children` (up to 16). Both the legacy
   * (`child`, `fork`) and N-way (`children`, `forks`) fields are returned.
   */
  async fork(params: ForkRequest): Promise<ForkResult> {
    if (!params.childCode && !params.children) {
      throw new Error("fork: provide either childCode or children");
    }
    if (params.childCode && params.children) {
      throw new Error("fork: childCode and children are mutually exclusive");
    }
    const body: Record<string, unknown> = {
      parent_code: params.parentCode,
    };
    if (params.childCode) body.child_code = params.childCode;
    if (params.children) {
      body.children = params.children.map((c) => {
        const child: Record<string, unknown> = { code: c.code };
        if (c.memoryMb !== undefined) child.memory_mb = c.memoryMb;
        return child;
      });
    }
    if (params.language)    body.language       = params.language;
    if (params.forkAfterMs !== undefined) body.fork_after_ms = params.forkAfterMs;
    if (params.timeoutSecs  !== undefined) body.timeout_secs  = params.timeoutSecs;
    if (params.memoryMb     !== undefined) body.memory_mb     = params.memoryMb;
    encodeExecOptions(body, params);
    return this.http.request<ForkResult>({
      method: "POST",
      path: "/v1/runs/fork",
      body,
    });
  }

  /**
   * Stream stdout/stderr lines as they are produced, via Server-Sent Events.
   *
   * Usage:
   *   for await (const ev of aj.runs.stream({ code, language })) {
   *     if (ev.type === "stdout") ...
   *   }
   */
  async *stream(params: RunRequest): AsyncGenerator<StreamEvent> {
    const body: Record<string, unknown> = { code: params.code };
    if (params.language)    body.language    = params.language;
    if (params.timeoutSecs !== undefined) body.timeout_secs = params.timeoutSecs;
    if (params.memoryMb    !== undefined) body.memory_mb    = params.memoryMb;
    encodeExecOptions(body, params);

    const res = await this.http.rawFetch({
      method: "POST",
      path: "/v1/runs/stream",
      headers: { accept: "text/event-stream", "content-type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!res.ok || !res.body) {
      const text = await res.text().catch(() => "");
      let parsed: unknown = null;
      try { parsed = text ? JSON.parse(text) : null; } catch { /* not JSON */ }
      throw new AgentjailError(
        res.status,
        parsed,
        text || res.statusText || "stream failed",
      );
    }

    yield* parseSSE(res.body);
  }
}

/**
 * Parse a `text/event-stream` into a stream of typed StreamEvents.
 * Tiny parser — doesn't try to handle every SSE edge case, just the shapes
 * our server actually emits (single `event:` + single `data:` per frame).
 */
async function* parseSSE(body: ReadableStream<Uint8Array>): AsyncGenerator<StreamEvent> {
  const decoder = new TextDecoder();
  const reader = body.getReader();
  let buffer = "";
  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      // SSE frames are separated by a blank line.
      let sep: number;
      while ((sep = buffer.indexOf("\n\n")) >= 0) {
        const frame = buffer.slice(0, sep);
        buffer = buffer.slice(sep + 2);
        const parsed = parseFrame(frame);
        if (parsed) yield parsed;
      }
    }
  } finally {
    reader.releaseLock();
  }
}

function parseFrame(frame: string): StreamEvent | null {
  let type = "";
  let data = "";
  for (const raw of frame.split("\n")) {
    if (raw.startsWith(":"))             continue;     // comment / keep-alive
    else if (raw.startsWith("event:"))   type = raw.slice(6).trim();
    else if (raw.startsWith("data:"))    data += raw.slice(5).trimStart();
  }
  if (!type) return null;

  switch (type) {
    case "stdout":   return { type: "stdout", line: data };
    case "stderr":   return { type: "stderr", line: data };
    case "started":
    case "stats":
    case "completed":
    case "error": {
      try {
        const payload = JSON.parse(data);
        return { type, ...payload } as StreamEvent;
      } catch {
        return null;
      }
    }
    default: return null;
  }
}

File: packages/sdk-python/src/agentjail/runs.py
"""Runs — one-shot code execution + fork + streaming."""

from __future__ import annotations

import json as _json
from collections.abc import Iterator
from typing import Any

from ._http import HttpClient
from .types import ExecResult, ForkChild, ForkResult, StreamEvent


class Runs:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(
        self,
        *,
        code: str,
        language: str | None = None,
        timeout_secs: int | None = None,
        memory_mb: int | None = None,
        network: Any = None,
        seccomp: str | None = None,
        cpu_percent: int | None = None,
        max_pids: int | None = None,
    ) -> ExecResult:
        body: dict[str, Any] = {"code": code}
        if language is not None:
            body["language"] = language
        if timeout_secs is not None:
            body["timeout_secs"] = timeout_secs
        if memory_mb is not None:
            body["memory_mb"] = memory_mb
        if network is not None:
            body["network"] = network
        if seccomp is not None:
            body["seccomp"] = seccomp
        if cpu_percent is not None:
            body["cpu_percent"] = cpu_percent
        if max_pids is not None:
            body["max_pids"] = max_pids
        return self._http.request("POST", "/v1/runs", json=body)

    def fork(
        self,
        *,
        parent_code: str,
        child_code: str | None = None,
        children: list[ForkChild] | None = None,
        language: str | None = None,
        fork_after_ms: int | None = None,
        timeout_secs: int | None = None,
        memory_mb: int | None = None,
        network: Any = None,
        seccomp: str | None = None,
        cpu_percent: int | None = None,
        max_pids: int | None = None,
    ) -> ForkResult:
        if not child_code and not children:
            raise ValueError("fork: provide either child_code or children")
        if child_code and children:
            raise ValueError("fork: child_code and children are mutually exclusive")
        body: dict[str, Any] = {"parent_code": parent_code}
        if child_code:
            body["child_code"] = child_code
        if children:
            body["children"] = children
        if language is not None:
            body["language"] = language
        if fork_after_ms is not None:
            body["fork_after_ms"] = fork_after_ms
        if timeout_secs is not None:
            body["timeout_secs"] = timeout_secs
        if memory_mb is not None:
            body["memory_mb"] = memory_mb
        if network is not None:
            body["network"] = network
        if seccomp is not None:
            body["seccomp"] = seccomp
        if cpu_percent is not None:
            body["cpu_percent"] = cpu_percent
        if max_pids is not None:
            body["max_pids"] = max_pids
        return self._http.request("POST", "/v1/runs/fork", json=body)

    def stream(
        self,
        *,
        code: str,
        language: str | None = None,
        timeout_secs: int | None = None,
        memory_mb: int | None = None,
    ) -> Iterator[StreamEvent]:
        """Run code and yield SSE events until the server closes the stream.

        Usage:

            for ev in aj.runs.stream(code="print(1)", language="python"):
                if ev["type"] == "stdout":
                    print(ev["line"])
        """
        body: dict[str, Any] = {"code": code}
        if language is not None:
            body["language"] = language
        if timeout_secs is not None:
            body["timeout_secs"] = timeout_secs
        if memory_mb is not None:
            body["memory_mb"] = memory_mb
        resp = self._http.stream("POST", "/v1/runs/stream", json=body)
        try:
            yield from _parse_sse(resp.iter_text())
        finally:
            resp.close()


def _parse_sse(chunks: Iterator[str]) -> Iterator[StreamEvent]:
    buffer = ""
    for chunk in chunks:
        buffer += chunk
        while True:
            sep = buffer.find("\n\n")
            if sep < 0:
                break
            frame = buffer[:sep]
            buffer = buffer[sep + 2 :]
            parsed = _parse_frame(frame)
            if parsed is not None:
                yield parsed


def _parse_frame(frame: str) -> StreamEvent | None:
    event_type = ""
    data = ""
    for raw in frame.split("\n"):
        if raw.startswith(":"):
            continue
        if raw.startswith("event:"):
            event_type = raw[6:].strip()
        elif raw.startswith("data:"):
            data += raw[5:].lstrip()
    if not event_type:
        return None
    if event_type in {"stdout", "stderr"}:
        return {"type": event_type, "line": data}  # type: ignore[return-value]
    if event_type in {"started", "stats", "completed", "error"}:
        try:
            payload = _json.loads(data)
        except ValueError:
            return None
        return {"type": event_type, **payload}  # type: ignore[return-value]
    return None
